fastestimator/estimator/trace.py

```diff
@@ -195,6 +195,97 @@ def on_epoch_end(self, state):
+        self.lr_schedule = lr_schedule
+        self.reduce_on_eval = reduce_on_eval
+        self.reduce_patience = reduce_patience
+        self.reduce_factor = reduce_factor
+        self.reduce_mode = reduce_mode
+        self.min_lr = min_lr
+        self.reduce_lr_ratio = 1.0
+        self.base_lr = None
+        self.current_lr = None
+        self.log_steps = None
+        self.model = None
+        self.change_lr = False
+        self.wait = 0
+        if self.lr_schedule:
+            assert isinstance(self.lr_schedule, LRSchedule), "lr_schedule must be instance of LRSchedule"
+        if self.reduce_mode == "min":
+            self.reduce_metric_best = np.Inf
+        elif self.reduce_mode == "max":
+            self.reduce_metric_best = -np.Inf
+        else:
+            raise ValueError("reduce_mode must be either 'min' or 'max'")
+
+    def on_begin(self, state):
+        self.model = self.network.model[self.model_name]
+        self.base_lr = backend.get_value(self.model.optimizer.lr)
+        self.current_lr = max(self.base_lr * self.reduce_lr_ratio, self.min_lr)
+        if self.reduce_on_eval is True:
+            self.reduce_on_eval = self.model.loss_name
+        if self.lr_schedule:
+            self.lr_schedule.initial_lr = self.current_lr
+
+    def on_epoch_begin(self, state):
+        if self.lr_schedule and self.lr_schedule.schedule_mode == "epoch":
+            self.base_lr = self.lr_schedule.schedule_fn(state["epoch"], self.base_lr)
+            self.change_lr = True
+
+    def on_batch_begin(self, state):
+        if state["mode"] == "train":
+            if self.lr_schedule and self.lr_schedule.schedule_mode == "step":
+                self.base_lr = self.lr_schedule.schedule_fn(state["train_step"], self.base_lr)
+                self.change_lr = True
+            if self.change_lr:
+                self._update_lr()
+
+    def on_batch_end(self, state):
+        if state["mode"] == "train" and self.log_steps and state["train_step"] % self.log_steps == 0:
+            state[self.model_name + "_lr"] = round(self.current_lr, 6)
+
+    def on_epoch_end(self, state):
+        if state["mode"] == "eval" and self.reduce_on_eval:
+            current_value = state[self.reduce_on_eval]
+            if self._is_better(current_value):
+                self.reduce_metric_best = current_value
+                self.wait = 0
+            else:
+                self.wait += 1
+                if self.wait >= self.reduce_patience:
+                    self.reduce_lr_ratio *= self.reduce_factor
+                    self.change_lr = True
+                    print("FastEstimator-LRController: learning rate reduced by factor of {}".format(
+                        self.reduce_factor))
+
+    def _update_lr(self):
+        self.current_lr = max(self.base_lr * self.reduce_lr_ratio, self.min_lr)
+        backend.set_value(self.model.optimizer.lr, self.current_lr)
+        self.change_lr = False
+
+    def _is_better(self, value):
```

Comment:
this could be done by selecting the comparison op up front when you're setting reduce_metric_best in the init method. see the early stopping trace for example

Reply:
done